src/Distribution/Server/Features/PackageRank.hs

```diff
@@ -0,0 +1,289 @@
+
+fracScor :: Float -> Float -> Scorer
+fracScor maxim frac = scorer maxim (maxim * frac)
+
+boolScor :: Float -> Bool -> Scorer
+boolScor k True  = Scorer k k
+boolScor k False = Scorer k 0
+
+total :: Scorer -> Float
+total (Scorer a b) = b / a
+
+major :: Num a => [a] -> a
+major (x : _) = x
+major _       = 0
+minor :: Num a => [a] -> a
+minor (_ : y : _) = y
+minor _           = 0
+patches :: Num a => [a] -> a
+patches (_ : _ : xs) = sum xs
+patches _            = 0
+
+numDays :: Maybe CL.UTCTime -> Maybe CL.UTCTime -> Float
+numDays (Just first) (Just end) =
+  fromRational $ toRational $ CL.diffUTCTime first end / fromRational
+    (toRational CL.nominalDay)
+numDays _ _ = 0
+
+freshness :: [Version] -> CL.UTCTime -> Bool -> IO Float
+freshness [] _ _ = return 0
+freshness (x : xs) lastUpd app =
+  daysPastExpiration
+    >>= (\dExp -> return $ max 0 $ (decayDays - dExp) / decayDays)
+ where
+  versionLatest = versionNumbers x
+  daysPastExpiration =
+    age >>= (\a -> return $ max 0 a - expectedUpdateInterval)
+  expectedUpdateInterval =
+    int2Float (min (versionStabilityInterval versionLatest) $ length (x : xs))
+  versionStabilityInterval v | patches v > 3 && major v > 0 = 700
+                             | patches v > 3                = 450
+                             | patches v > 0                = 300
+                             | major v > 0                  = 200
+                             | minor v > 3                  = 140
+                             | otherwise                    = 80
+  age       = flip numDays (Just lastUpd) . Just <$> CL.getCurrentTime
+  decayDays = expectedUpdateInterval / 2 + (if app then 300 else 200)
+
+-- lookupPackageId
+-- queryHasDocumentation
+
+-- TODO CoreFeature can be substituted by CoreResource
+rankIO
+  :: VersionsFeature
+  -> Int
+  -> Int
+  -> DocumentationFeature
+  -> ServerEnv
+  -> TarIndexCacheFeature
+  -> [PkgInfo]
+  -> Maybe PkgInfo
+  -> IO Scorer
+
+rankIO _ _ _ _ _ _ _ Nothing = return (Scorer (118 + 16 + 4 + 1) 0)
+rankIO vers recentDownloads maintainers docs env tarCache pkgs (Just pkgI) = do
+  temp  <- temporalScore pkg lastUploads versionList recentDownloads
+  versS <- versionScore versionList vers lastUploads pkg
+  codeS <- codeScore documentLines srcLines
+  return (temp <> versS <> codeS <> authorScore maintainers pkg)
+
+ where
+  pkg   = packageDescription $ pkgDesc pkgI
+  pkgId = package pkg
+  lastUploads =
+    sortBy (flip compare)
+      $  (fst . pkgOriginalUploadInfo <$> pkgs)
+      ++ (fst . pkgLatestUploadInfo <$> pkgs)
+  versionList :: [Version]
+  versionList = sortBy (flip compare)
+    $ map (pkgVersion . package . packageDescription) (pkgDesc <$> pkgs)
+  packageEntr = do
+    tarB <- packageTarball tarCache pkgI
+    return
+      $   (\(path, _, index) -> (path, ) <$> T.lookup index path)
+      =<< rightToMaybe tarB
+  rightToMaybe (Right a) = Just a
+  rightToMaybe (Left  _) = Nothing
```

Comment:
Can't you take advantage of the fact that there is MonadFail Maybe? It also doesn't make sense to me to mix do-notation and calls to bind (`=<<`). I'd almost always prefer do notation since it is more readable to have identifiers.

I think the code is doing something like this, which I think is a lot simpler:

```
packageEntr = do
    Right (path, _, index) <- packageTarball tarCache pkgI
    foobar <- T.lookup index path
    pure (path, foobar)
```

Reply:
This part of the code does not actually work so when I fix it I will format it nicer. I was not aware of MonadFail this will probably make a lot of things simpler.